graphql_compiler/schema_transformation/split_query.py

```diff
@@ -0,0 +1,642 @@
+    return ast_types.Document(
+        definitions=[
+            ast_types.OperationDefinition(
+                operation='query',
+                selection_set=ast_types.SelectionSet(
+                    selections=[
+                        ast_types.Field(
+                            name=ast_types.Name(value=root_vertex_field_name),
+                            # NOTE: if the root_vertex_field_name does not actually exist
+                            # as a root field (not all types are required to have a
+                            # corresponding root vertex field), then this query will be
+                            # invalid
+                            # TODO: warn the user?
+                            selection_set=ast_types.SelectionSet(
+                                selections=root_selections,
+                            ),
+                            directives=[],
+                        )
+                    ]
+                )
+            )
+        ]
+    )
+
+
+def _get_output_directive(out_name):
+    """Return a Directive representing an @output with the input out_name."""
+    return ast_types.Directive(
+        name=ast_types.Name(value=u'output'),
+        arguments=[
+            ast_types.Argument(
+                name=ast_types.Name(value=u'out_name'),
+                value=ast_types.StringValue(value=out_name),
+            ),
+        ],
+    )
+
+
+def _get_out_name_optionally_add_output(field, intermediate_out_name_assigner):
+    """Return out_name of @output on field, creating new @output if needed.
+
+    Args:
+        field: Field object, whose directives we may modify by adding an @output directive
+        intermediate_out_name_assigner: IntermediateOutNameAssigner, which will be used to
+                                        generate an out_name, if it's necessary to create a
+                                        new @output directive
+
+    Returns:
+        str, name of the out_name of the @output directive, either pre-existing or newly
+        generated
+    """
+    # Check for existing directive
+    output_directive = try_get_ast(field.directives, u'output', ast_types.Directive)
+    if output_directive is None:
+        # Create and add new directive to field
+        out_name = intermediate_out_name_assigner.assign_and_return_out_name()
+        output_directive = _get_output_directive(out_name)
+        if field.directives is None:
+            field.directives = []
+        field.directives.append(output_directive)
+        return out_name
+    else:
+        return output_directive.arguments[0].value.value  # Location of value of out_name
+
+
+class IntermediateOutNameAssigner(object):
+    """Used to generate and keep track of out_name of @output directives."""
+    def __init__(self):
+        """Create assigner with empty records."""
+        self.intermediate_output_names = set()
+        self.intermediate_output_count = 0
```

Comment:
Could we replace this with `len(self.intermediate_output_names)` wherever it's used, or is it possible for this field to diverge from the length of the names set?

Reply:
My feeling is that it doesn't matter a that much how this is represented. This is a tentative way of generating intermediate output names. I made this object to assign and keep track of names, so that it's easy to change. In its currently iteration, the size of the set is equal to the count, but this could easily change I think.

Reply:
My feeling is that it doesn't matter a that much how this is represented. This is a tentative way of generating intermediate output names. I made this object to assign and keep track of names, so that it's easy to change. In its currently iteration, the size of the set is equal to the count, but this could easily change I think.